core/consensus/common/src/import_queue/basic_queue.rs

```diff
@@ -361,9 +284,14 @@ fn import_many_blocks<B: BlockT, V: Verifier<B>>(
 
 	// Blocks in the response/drain should be in ascending order.
 	for block in blocks {
+		if !keep_going() {
+			return (imported, count, results);
+		}
```

Comment:
I'm wondering if this mechanism could change the logic of the queue, and whether that's intended. 

Previously, if there was an error in any block, we would continue to loop over the entire batch and mark all as errored, then send back results for the entire batch. 

This would then ensure all hashes are deleted from sync(I guess otherwise sync thinks the blocks are still processed by the queue?).

If you were to set `has_error` to true when `keep_going` returns false, I think you could keep the logic the same. 

It might not matter if `keep_going` being false mean the entire system is shutting down, but I guess it wouldn't hurt.